### PR TITLE
Add valid spark on k8s airflow connection urls to tests

### PR DESCRIPTION
**Pull request by amoghrajesh**
Airflow connections parses valid urls fine. The issue described in #21638 provided a wrong url as input to the connections with double `://` leading to parsing issues. This PR simply adds a few valid spark on k8s connection urls to the existing test suite. 

Testing by adding various different unit test cases to cover the breadth

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/models/connection.py
Comment (uranusjr) [inline airflow/models/connection.py]: Why do you need to do this, `uri` is already a `SplitResult` instance…?

**Comment by uranusjr** [comment]
> Airflow connections is unable to parse the uri strings that contain another level of scheme in their uri strings. For example: `spark://k8s://100.68.0.1:443?deploy-mode=cluster`.

This is not a valid URI to begin with. You should escape the second `://` instead.

File: airflow/models/connection.py
Comment (amoghrajesh) [reply airflow/models/connection.py]: Thanks for your comment. No, one line 200 we are doing a list comprehension. This is leading to the type being as `list`.

Verified by adding a few log statements:
```
def split_uri_to_parts(uri):
    c = uri.count("://")
    if c < 2:
        return urlsplit(uri), False
    idx = uri.rfind("://")
    # replace :// with :__
    uri_list = list(uri)
    uri_list[idx : idx + 3] = ":__"
    uri = "".join(uri_list)
    uri = urlsplit(uri)
    uri = [e for e in uri]
    # replace back :__ with ://
    uri[1] = uri[1].replace(":__", "://")

    print(type(uri))

    uri = SplitResult(uri[0], uri[1], uri[2], uri[3], uri[4])

    print(type(uri))
    return uri, True
```

Output is:
```
<class 'list'>
<class 'urllib.parse.SplitResult'>
(SplitResult(scheme='spark', netloc='k8s://100.68.0.1:443', path='', query='deploy-mode=cluster', fragment=''), True)
```

**Comment by amoghrajesh** [comment]
> > Airflow connections is unable to parse the uri strings that contain another level of scheme in their uri strings. For example: `spark://k8s://100.68.0.1:443?deploy-mode=cluster`.
> 
> This is not a valid URI to begin with. You should escape the second `://` instead.

Oh I wasn't aware of that. Made the changes accordingly

File: airflow/models/connection.py
Comment (uranusjr) [reply airflow/models/connection.py]: And why do you need to do that list comprehention in the first place? SplitResult supports index access out of the box.

Comment (amoghrajesh) [reply airflow/models/connection.py]: Hello, I am trying to re-assign a variable to SplitResult list. That list does not support reassignment. Hence to achieve something like that, I had to convert it to a list, re-assign the elements and then convert it back to the SplitResult list. @uranusjr 
```
/Users/adesai/.pyenv/versions/airflow-env/bin/python /Users/adesai/Documents/OpenSource/Airflow/airflow/airflow/models/test.py
Traceback (most recent call last):
  File "/Users/adesai/Documents/OpenSource/Airflow/airflow/airflow/models/test.py", line 24, in <module>
    parts = split_uri_to_parts("spark://k8s://100.68.0.1:443?deploy-mode=cluster")
  File "/Users/adesai/Documents/OpenSource/Airflow/airflow/airflow/models/test.py", line 15, in split_uri_to_parts
    uri[1] = uri[1].replace(":__", "://")
TypeError: 'SplitResult' object does not support item assignment
```

Comment (uranusjr) [reply airflow/models/connection.py]: You should use `_replace`. See documentation on namedtuple.

Comment (amoghrajesh) [reply airflow/models/connection.py]: Thank you! Checked it out and it is brilliant. Pushing a commit with the fix.

**Comment by amoghrajesh** [comment]
@uranusjr can you take a look at this PR once you have some time? I have addressed your review comments.

File: airflow/models/connection.py
Comment (uranusjr) [inline airflow/models/connection.py]: Is it ever possible for this to return >=2 if the escape syntax is used…? A valid URL can never contain more than one `://` sequence.

Comment (amoghrajesh) [reply airflow/models/connection.py]: Good point, since `spark://k8s://100.68.0.1:443?deploy-mode=cluster` is not a valid URL as you mentioned earlier. I think the logic in this PR is to have a fallback mechanism in case such a scenario occurs.

Comment (amoghrajesh) [reply airflow/models/connection.py]: In normal cases where the format is fine, the logic minus what this PR adds seems to work fine. I think we can do just by adding the cases we have in `test_connection.py` as far as this PR is concerned. What do you think @uranusjr ?

Comment (uranusjr) [reply airflow/models/connection.py]: Yeah makes sense. Also could be a good idea to add something in the Connection form to emit an error if the user puts in an invalud URL. Probably cleaner to do in a separate PR though.

Comment (amoghrajesh) [reply airflow/models/connection.py]: @uranusjr do you mean on the UI? I am not sure if we have a check for that, it would be great to do that though.

Comment (amoghrajesh) [reply airflow/models/connection.py]: As far as this PR is concerned, I will undo all the changes and just add the test cases, that way we have confidence that the actual code for this works.

Comment (uranusjr) [reply airflow/models/connection.py]: I think this can be done in `ConnectionForm` (see WTForms documentation to add validators). We should probably do that on the `host` field.

Comment (amoghrajesh) [reply airflow/models/connection.py]: Nice. Let me come up with a separate issue and PR for that :)

**Review by uranusjr — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by amoghrajesh** [comment]
@potiuk @uranusjr can you help in merging this PR? It has passed the checks

**Comment by eladkal** [comment]
This PR marked as closing for https://github.com/apache/airflow/issues/21638 but it changes only the test.
How can this fix a user bug when just adding more test coverage without fixing anything in the code itself?

**Comment by amoghrajesh** [comment]
@eladkal the issue described in https://github.com/apache/airflow/issues/21638 is not a common use case. This PR adds some more scenarios to the test case, including what is mentioned in the github issue. Maybe we should not mark is as `closes` in that case. WDYT @eladkal?
Also, new issue to fix this properly is up https://github.com/apache/airflow/issues/31460

**Comment by eladkal** [comment]
> This PR adds some more scenarios to the test case, including what is mentioned in the github issue. 

that is OK just don't mark the PR as closing the issue since it does not

**Comment by amoghrajesh** [comment]
Removed that portion from the description

**Comment by amoghrajesh** [comment]
@uranusjr @potiuk can we merge this PR? Clean build and 2 approvals so far.

**Comment by hussein-awala** [comment]
@amoghrajesh, I'm currently working on the fix. Let's wait before merging this PR to see if we need to add more tests.

**Comment by amoghrajesh** [comment]
@hussein-awala a related issue that might be interesting or help you with the fix: #31460

**Comment by amoghrajesh** [comment]
@hussein-awala can we merge this one in after we merge the dependent PR?

**Comment by amoghrajesh** [comment]
@hussein-awala can we merge this PR too as the dependent PR has been merged recently?

**Pull request merged**
